File: Sources/AIKit/AnthropicProvider.swift
import Foundation

// MARK: - AnthropicProvider

/// `AIProvider` for Anthropic's Messages API (`POST /v1/messages`).
///
/// Anthropic does not allow "system" role in the messages array — it must be
/// a top-level "system" field. This provider filters system messages out of
/// the array and uses the last system message's content as the top-level field.
///
/// Supports multipart messages: text parts become `{"type":"text","text":"..."}`,
/// image parts become `{"type":"image","source":{"type":"base64",...}}`.
/// System messages are always text-only (Anthropic API constraint).
public final class AnthropicProvider: AIProvider, Sendable {

    public typealias HTTPClient = @Sendable (URLRequest) async throws -> (Data, URLResponse)

    // MARK: Stored

    private let apiKey: String
    private let endpoint: URL
    private let httpClient: HTTPClient

    // MARK: Init

    public init(
        apiKey: String,
        endpoint: URL = AnthropicProvider.defaultEndpoint,
        httpClient: @escaping HTTPClient = AnthropicProvider.defaultHTTPClient
    ) {
        self.apiKey = apiKey
        self.endpoint = endpoint
        self.httpClient = httpClient
    }

    // MARK: Defaults

    public static let defaultEndpoint = URL(string: "https://api.anthropic.com/v1/messages")!
    public static let defaultModel = "claude-sonnet-4-6"

    public static let defaultHTTPClient: HTTPClient = { request in
        try await URLSession.shared.data(for: request)
    }

    // MARK: AIProvider

    public func chat(messages: [ChatMessage], config: AIConfig) async throws -> String {
        let request = try Self.buildRequest(
            endpoint: endpoint,
            apiKey: apiKey,
            messages: messages,
            config: config
        )

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await httpClient(request)
        } catch {
            throw AIError.sendFailed(message: error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw AIError.sendFailed(message: "Non-HTTP response from Anthropic API")
        }

        switch httpResponse.statusCode {
        case 200:
            return try Self.parse(data: data)
        case 401:
            throw AIError.authenticationFailed
        case 429:
            throw AIError.rateLimited
        default:
            let body = String(data: data, encoding: .utf8) ?? "<unreadable body>"
            throw AIError.sendFailed(message: "HTTP \(httpResponse.statusCode): \(body)")
        }
    }

    // MARK: - Request builder (internal for tests)

    static func buildRequest(
        endpoint: URL,
        apiKey: String,
        messages: [ChatMessage],
        config: AIConfig
    ) throws -> URLRequest {
        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")
        request.setValue("2023-06-01", forHTTPHeaderField: "anthropic-version")
        request.setValue("application/json", forHTTPHeaderField: "content-type")

        // Separate system messages from conversation messages.
        // Anthropic rejects "system" role in the messages array;
        // use the last system message's content as the top-level field.
        let systemMessages = messages.filter { $0.role == .system }
        let conversationMessages = messages.filter { $0.role != .system }

        // Prefer explicit config.systemPrompt; fall back to last system message.
        let systemField: String? = config.systemPrompt ?? systemMessages.last?.text

        var body: [String: Any] = [
            "model": config.model,
            "max_tokens": config.maxTokens,
            "temperature": config.temperature,
            "messages": conversationMessages.map { msg -> [String: Any] in
                ["role": msg.role.rawValue, "content": Self.serializeParts(msg.parts)]
            },
        ]
        if let system = systemField {
            body["system"] = system
        }

        do {
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        } catch {
            throw AIError.sendFailed(message: "Could not serialize request body: \(error.localizedDescription)")
        }
        return request
    }

    // MARK: - Response parser (internal for tests)

    static func parse(data: Data) throws -> String {
        struct Response: Decodable {
            struct ContentBlock: Decodable {
                let type: String
                let text: String?
            }
            let content: [ContentBlock]
        }

        let response: Response
        do {
            response = try JSONDecoder().decode(Response.self, from: data)
        } catch {
            throw AIError.decodingFailed(message: error.localizedDescription)
        }

        // Concatenate all text-type blocks in order.
        let text = response.content
            .filter { $0.type == "text" }
            .compactMap { $0.text }
            .joined()

        return text
    }

    // MARK: - Private: part serialization

    /// Convert structured message parts to Anthropic content-block JSON objects.
    /// text → {"type":"text","text":"..."}
    /// image → {"type":"image","source":{"type":"base64","media_type":"image/jpeg","data":"<b64>"}}
    private static func serializeParts(_ parts: [ChatMessage.Part]) -> Any {
        // Single text-only part: send as plain string for maximum API compatibility.
        if parts.count == 1, case .text(let s) = parts[0] {
            return s
        }
        return parts.map { part -> [String: Any] in
            switch part {
            case .text(let s):
                return ["type": "text", "text": s]
            case .image(let jpegData, let mimeType):
                return [
                    "type": "image",
                    "source": [
                        "type": "base64",
                        "media_type": mimeType,
                        "data": jpegData.base64EncodedString(),
                    ] as [String: Any],
                ]
            }
        }
    }
}

File: windows/src/KosmoNotes.Providers/AnthropicProvider.cs
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;
using KosmoNotes.Core.Models;
using KosmoNotes.Providers.Internal;

namespace KosmoNotes.Providers;

/// <summary>
/// <see cref="IAIProvider"/> for Anthropic's Messages API
/// (<c>POST https://api.anthropic.com/v1/messages</c>). Mirrors the Swift
/// <c>AnthropicProvider</c> exactly.
/// </summary>
/// <remarks>
/// Anthropic does not allow <c>"system"</c> role inside the messages array —
/// it must be a top-level <c>"system"</c> field. This provider filters
/// system messages out of the array and uses the last system message's
/// content as the top-level field, unless <see cref="AIConfig.SystemPrompt"/>
/// is set (which takes priority).
///
/// Multipart messages: text parts become <c>{"type":"text","text":"..."}</c>;
/// image parts become <c>{"type":"image","source":{"type":"base64",...}}</c>.
/// A message with a single text part is sent as a plain string, which is
/// what the API expects in the common case.
/// </remarks>
public sealed class AnthropicProvider : IAIProvider
{
    /// <summary>The API endpoint URL when no override is supplied.</summary>
    public static readonly Uri DefaultEndpoint =
        new("https://api.anthropic.com/v1/messages");

    /// <summary>The model identifier used by default in <see cref="AIConfig"/>.</summary>
    public const string DefaultModel = "claude-sonnet-4-6";

    private readonly string _apiKey;
    private readonly Uri _endpoint;
    private readonly HttpClient _httpClient;

    /// <summary>Create a provider using the given API key.</summary>
    /// <param name="apiKey">Anthropic API key (sent as <c>x-api-key</c> header).</param>
    /// <param name="endpoint">
    /// Optional endpoint override. Defaults to <see cref="DefaultEndpoint"/>.
    /// </param>
    /// <param name="httpClient">
    /// Optional client for tests; uses a process-wide default when null.
    /// </param>
    public AnthropicProvider(string apiKey, Uri? endpoint = null, HttpClient? httpClient = null)
    {
        _apiKey = apiKey;
        _endpoint = endpoint ?? DefaultEndpoint;
        _httpClient = httpClient ?? HttpDefaults.Default;
    }

    /// <inheritdoc />
    public async Task<string> ChatAsync(
        IReadOnlyList<ChatMessage> messages,
        AIConfig config,
        CancellationToken ct = default)
    {
        using var request = BuildRequest(_endpoint, _apiKey, messages, config);

        HttpResponseMessage response;
        try
        {
            response = await _httpClient.SendAsync(request, ct).ConfigureAwait(false);
        }
        catch (HttpRequestException ex)
        {
            throw new SendFailedException(ex.Message, ex);
        }

        try
        {
            await HttpStatus.ThrowIfNotSuccessAsync(response, ct).ConfigureAwait(false);
            byte[] body = await response.Content.ReadAsByteArrayAsync(ct).ConfigureAwait(false);
            return Parse(body);
        }
        finally
        {
            response.Dispose();
        }
    }

    /// <summary>
    /// Build the <see cref="HttpRequestMessage"/> for a chat call. Internal so
    /// tests can inspect the wire shape without making a real HTTP call.
    /// </summary>
    internal static HttpRequestMessage BuildRequest(
        Uri endpoint,
        string apiKey,
        IReadOnlyList<ChatMessage> messages,
        AIConfig config)
    {
        var request = new HttpRequestMessage(HttpMethod.Post, endpoint);
        request.Headers.TryAddWithoutValidation("x-api-key", apiKey);
        request.Headers.TryAddWithoutValidation("anthropic-version", "2023-06-01");

        // Separate system messages from conversation messages.
        // Anthropic rejects "system" role in the messages array;
        // use the last system message's content as the top-level field.
        var conversationMessages = messages.Where(m => m.Role != ChatRole.System).ToList();
        ChatMessage? lastSystem = messages.LastOrDefault(m => m.Role == ChatRole.System);

        // Prefer explicit config.SystemPrompt; fall back to last system message text.
        string? systemField = config.SystemPrompt ?? lastSystem?.Text;

        var body = new JsonObject
        {
            ["model"] = config.Model,
            ["max_tokens"] = config.MaxTokens,
            ["temperature"] = config.Temperature,
            ["messages"] = SerializeMessages(conversationMessages),
        };
        if (systemField is not null)
        {
            body["system"] = systemField;
        }

        string json = body.ToJsonString();
        request.Content = new StringContent(json, Encoding.UTF8);
        request.Content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
        return request;
    }

    /// <summary>Parse a 200-response body and concatenate all text blocks.</summary>
    internal static string Parse(byte[] data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            var root = doc.RootElement;
            if (!root.TryGetProperty("content", out var content) ||
                content.ValueKind != JsonValueKind.Array)
            {
                throw new DecodingFailedException("Missing 'content' array.");
            }

            var sb = new StringBuilder();
            foreach (var block in content.EnumerateArray())
            {
                if (block.TryGetProperty("type", out var type) &&
                    type.GetString() == "text" &&
                    block.TryGetProperty("text", out var text))
                {
                    sb.Append(text.GetString());
                }
            }
            return sb.ToString();
        }
        catch (DecodingFailedException)
        {
            throw;
        }
        catch (JsonException ex)
        {
            throw new DecodingFailedException(ex.Message, ex);
        }
    }

    private static JsonArray SerializeMessages(IEnumerable<ChatMessage> messages)
    {
        var arr = new JsonArray();
        foreach (var msg in messages)
        {
            arr.Add(new JsonObject
            {
                ["role"] = RoleToString(msg.Role),
                ["content"] = SerializeParts(msg.Parts),
            });
        }
        return arr;
    }

    private static JsonNode SerializeParts(IReadOnlyList<ChatPart> parts)
    {
        // Single text-only part: send as plain string for max API compatibility.
        if (parts.Count == 1 && parts[0] is TextPart only)
        {
            return JsonValue.Create(only.Text)!;
        }

        var arr = new JsonArray();
        foreach (var part in parts)
        {
            switch (part)
            {
                case TextPart t:
                    arr.Add(new JsonObject { ["type"] = "text", ["text"] = t.Text });
                    break;
                case ImagePart img:
                    arr.Add(new JsonObject
                    {
                        ["type"] = "image",
                        ["source"] = new JsonObject
                        {
                            ["type"] = "base64",
                            ["media_type"] = img.MimeType,
                            ["data"] = Convert.ToBase64String(img.JpegData),
                        },
                    });
                    break;
            }
        }
        return arr;
    }

    private static string RoleToString(ChatRole role) => role switch
    {
        ChatRole.System => "system",
        ChatRole.User => "user",
        ChatRole.Assistant => "assistant",
        _ => throw new ArgumentOutOfRangeException(nameof(role)),
    };
}
